File: metacasa-ios/MetaCasa/Models/Transaction.swift
import Foundation

struct Transaction: Codable, Identifiable, Hashable, Sendable {
    let id: UUID
    var householdId: UUID
    var userId: UUID
    var accountId: UUID?
    var type: TxType
    var amount: Decimal
    var amountOriginal: Decimal?
    var currencyOriginal: String?
    var fxRateToBase: Decimal?
    var fxSource: String?
    var fxStatus: String?
    var category: String
    var subcategory: String?
    var account: String?
    var note: String?
    var date: Date
    var periodYear: Int?
    var periodMonth: Int?

    /// Vincula las dos piernas de una transferencia entre cuentas propias. `nil` = movimiento normal.
    ///
    /// REGLA (ver `20260803140000_transfer_group_id.sql`):
    /// - Los agregados de ingreso/gasto/categoría/score DEBEN excluirlas: mover plata entre cuentas
    ///   propias no es ni ingreso ni gasto del hogar, la plata sigue adentro. Sumarlas infla ambos
    ///   lados a la vez y hunde el health score.
    /// - Los agregados de SALDO POR CUENTA **no** las excluyen: ahí las dos piernas SON el mecanismo
    ///   por el que la plata se mueve de una cuenta a la otra.
    ///
    /// Usar el helper `[Transaction].excludingTransfers` en vez de escribir el filtro a mano.
    var transferGroupId: UUID?

    let createdAt: Date?

    enum CodingKeys: String, CodingKey {
        case id
        case householdId = "household_id"
        case userId = "user_id"
        case accountId = "account_id"
        case type
        case amount
        case amountOriginal = "amount_original"
        case currencyOriginal = "currency_original"
        case fxRateToBase = "fx_rate_to_base"
        case fxSource = "fx_source"
        case fxStatus = "fx_status"
        case category
        case subcategory
        case account
        case note
        case date
        case periodYear = "period_year"
        case periodMonth = "period_month"
        case transferGroupId = "transfer_group_id"
        case createdAt = "created_at"
    }

    /// `true` si es una de las dos piernas de una transferencia entre cuentas propias.
    var isTransfer: Bool { transferGroupId != nil }
}

extension Array where Element == Transaction {
    /// Quita las transferencias entre cuentas propias.
    ///
    /// Existe como helper y no como filtro suelto para que el criterio sea buscable: cuando alguien
    /// agregue un agregado nuevo, `grep excludingTransfers` muestra todos los que ya lo aplican.
    /// El auditor contó ~35 sitios en iOS que suman dinero; escribir `$0.transferGroupId == nil` a
    /// mano en cada uno es cómo se olvida el número 36.
    var excludingTransfers: [Transaction] { filter { !$0.isTransfer } }
}

enum TxType: String, Codable, Hashable, Sendable, CaseIterable {
    case gasto = "GASTO"
    case ingreso = "INGRESO"

    var label: String {
        switch self {
        case .gasto: "Gasto"
        case .ingreso: "Ingreso"
        }
    }

    var systemIcon: String {
        switch self {
        case .gasto: "arrow.down.circle.fill"
        case .ingreso: "arrow.up.circle.fill"
        }
    }
}

/// Payload para insertar una tx nueva. Supabase completará id/created_at.
struct NewTransactionInput: Codable, Sendable {
    var householdId: UUID
    var userId: UUID
    var accountId: UUID?
    var type: TxType
    /// **Siempre en la moneda BASE del hogar.** Es el contrato canónico, escrito en
    /// `metacasa-web/AGENTS_CONTRACT.md`: todos los agregados (totales del Home,
    /// `envelope_balance`, el matview mensual) asumen que esta columna está en base.
    var amount: Decimal
    /// Monto tal como lo tipeó el usuario, en `currencyOriginal`.
    ///
    /// Faltaba, y por eso el original se perdía para siempre: iOS guardaba el monto ya convertido
    /// pero lo etiquetaba con la moneda extranjera. Un gasto de USD 100 en un hogar ARS a 1500 se
    /// guardaba como 150000 con `currency_original = "USD"`, y la lista mostraba "US$ 150.000".
    var amountOriginal: Decimal?
    var currencyOriginal: String?
    /// Cuántas unidades de la moneda base equivalen a 1 de `currencyOriginal`.
    /// Vale 1 cuando no hubo conversión. Invariante: `amount == amountOriginal * fxRateToBase`.
    var fxRateToBase: Decimal?
    var category: String
    var subcategory: String?
    var note: String?

    /// Fecha del movimiento, **siempre normalizada a mediodía UTC** por el init.
    ///
    /// El `private(set)` y el init son a propósito: la normalización tiene que ser imposible de
    /// saltear. Hay 8 caminos que crean transacciones (alta manual, import CSV, dos App Intents,
    /// asistente, recibos, voz, tool de IA) y parchear cada uno dejaba la puerta abierta al noveno.
    /// Poniéndolo acá, el que agregue un camino nuevo lo hereda sin enterarse.
    private(set) var date: Date

    init(
        householdId: UUID,
        userId: UUID,
        accountId: UUID?,
        type: TxType,
        amount: Decimal,
        amountOriginal: Decimal? = nil,
        currencyOriginal: String?,
        fxRateToBase: Decimal? = nil,
        category: String,
        subcategory: String?,
        note: String?,
        date: Date
    ) {
        self.householdId = householdId
        self.userId = userId
        self.accountId = accountId
        self.type = type
        self.amount = amount
        self.amountOriginal = amountOriginal
        self.currencyOriginal = currencyOriginal
        self.fxRateToBase = fxRateToBase
        self.category = category
        self.subcategory = subcategory
        self.note = note
        self.date = date.stableForStorage()
    }

    enum CodingKeys: String, CodingKey {
        case householdId = "household_id"
        case userId = "user_id"
        case accountId = "account_id"
        case type
        case amount
        case amountOriginal = "amount_original"
        case currencyOriginal = "currency_original"
        case fxRateToBase = "fx_rate_to_base"
        case category
        case subcategory
        case note
        case date
    }
}

File: metacasa-flutter/lib/models/transaction.dart
import 'package:decimal/decimal.dart';
import 'package:freezed_annotation/freezed_annotation.dart';

import '../core/utils/decimal_converter.dart';

part 'transaction.freezed.dart';
part 'transaction.g.dart';

/// Tipo de movimiento. Wire-values en MAYÚSCULAS (`GASTO`/`INGRESO`) — espejo
/// de `TxType` en iOS y del enum Postgres.
enum TxType {
  @JsonValue('GASTO')
  gasto,
  @JsonValue('INGRESO')
  ingreso;

  /// Etiqueta legible (es-AR).
  String get label => switch (this) {
        TxType.gasto => 'Gasto',
        TxType.ingreso => 'Ingreso',
      };
}

/// Movimiento (transacción). Espejo 1:1 de `public.transactions`.
@freezed
abstract class Transaction with _$Transaction {
  const factory Transaction({
    required String id,
    @JsonKey(name: 'household_id') required String householdId,
    @JsonKey(name: 'user_id') required String userId,
    @JsonKey(name: 'account_id') String? accountId,
    required TxType type,
    @DecimalConverter() required Decimal amount,
    @JsonKey(name: 'amount_original')
    @DecimalNullConverter()
    Decimal? amountOriginal,
    @JsonKey(name: 'currency_original') String? currencyOriginal,
    @JsonKey(name: 'fx_rate_to_base')
    @DecimalNullConverter()
    Decimal? fxRateToBase,
    @JsonKey(name: 'fx_source') String? fxSource,
    @JsonKey(name: 'fx_status') String? fxStatus,
    required String category,
    String? subcategory,
    String? account,
    String? note,
    required DateTime date,
    @JsonKey(name: 'period_year') int? periodYear,
    @JsonKey(name: 'period_month') int? periodMonth,
    /// Une las DOS piernas de una transferencia entre cuentas propias: un GASTO en
    /// la cuenta de origen y un INGRESO en la de destino, con el mismo id.
    ///
    /// Es una columna aditiva y no un valor nuevo de [TxType] a propósito. Un valor
    /// de enum desconocido hace fallar `$enumDecode` con `ArgumentError`, y como el
    /// decode es de lista, UNA fila rara vacía la pantalla entera en vez de mostrar
    /// un movimiento raro. Una key JSON extra, en cambio, la ignora cualquier
    /// cliente viejo que no la conozca.
    @JsonKey(name: 'transfer_group_id') String? transferGroupId,
    @JsonKey(name: 'created_at') DateTime? createdAt,
  }) = _Transaction;

  factory Transaction.fromJson(Map<String, dynamic> json) =>
      _$TransactionFromJson(json);
}

/// Payload para insertar una tx nueva. Supabase completa id/created_at.
/// Espejo de `NewTransactionInput` en iOS.
@freezed
abstract class NewTransactionInput with _$NewTransactionInput {
  const factory NewTransactionInput({
    @JsonKey(name: 'household_id') required String householdId,
    @JsonKey(name: 'user_id') required String userId,
    @JsonKey(name: 'account_id') String? accountId,
    required TxType type,
    /// **Siempre en la moneda BASE del hogar** — es el contrato de
    /// `metacasa-web/AGENTS_CONTRACT.md`: todos los agregados lo asumen.
    @DecimalConverter() required Decimal amount,
    /// Monto tal como lo tipeó el usuario, en [currencyOriginal].
    @JsonKey(name: 'amount_original')
    @DecimalNullConverter()
    Decimal? amountOriginal,
    @JsonKey(name: 'currency_original') String? currencyOriginal,
    /// Cuántas unidades de base equivalen a 1 de [currencyOriginal]. 1 si no hubo
    /// conversión. Invariante: `amount == amountOriginal * fxRateToBase`.
    @JsonKey(name: 'fx_rate_to_base')
    @DecimalNullConverter()
    Decimal? fxRateToBase,
    required String category,
    String? subcategory,
    String? note,
    required DateTime date,
  }) = _NewTransactionInput;

  factory NewTransactionInput.fromJson(Map<String, dynamic> json) =>
      _$NewTransactionInputFromJson(json);
}

/// La regla de las transferencias, con nombre.
///
/// Una transferencia entre cuentas propias son DOS movimientos con el mismo
/// [Transaction.transferGroupId]. Eso es lo que mueve bien los saldos por cuenta, pero la
/// plata nunca salió del hogar: **agregarle una transferencia a un conjunto de movimientos
/// no puede cambiar ningún total de ingreso, gasto ni categoría**.
///
/// Va como extensión con nombre y no como un `where(...)` suelto por lo mismo que
/// `txs.excludingTransfers` en iOS: con muchos sitios que suman dinero, escribir el
/// predicado a mano en cada uno es cómo se olvida el último. Con el nombre, un `grep`
/// muestra quién ya lo aplica.
///
/// **Dónde NO va:** en los saldos por cuenta las dos piernas son el mecanismo, no ruido;
/// filtrarlas ahí rompe el saldo, que es peor que el bug original. Tampoco en la lista de
/// movimientos: el usuario tiene que poder ver sus transferencias.
extension TransferFiltering on List<Transaction> {
  List<Transaction> get excludingTransfers =>
      where((Transaction t) => t.transferGroupId == null).toList();
}

extension TransferPredicate on Transaction {
  bool get isTransfer => transferGroupId != null;
}
